### PR TITLE
Wire up the builders for dev3x

### PR DESCRIPTION
**Pull request by MarkZuber**
This PR is to wire up the new ClientApp builders and the AcquireToken builders.  
It does not yet compile fully as there's more work to do, but since this is going into dev3x I'd prefer to get signoff and cleanup at this point before I hit the next stage of restructuring the Authority classes to use AuthorityInfo from the AppBuilders, removing the remaining statics and doing further [Obsolete] / migration work.  Much of this code I already have working in the earlier config branch so I have confidence in getting tests and functionality back.

### CODE REVIEW

File: src/Microsoft.Identity.Client/ClientApplicationBase.WithBuilders.cs [hunk 1]
```diff
@@ -25,14 +25,67 @@
...
+            ServiceBundle = Core.ServiceBundle.Create(config);
+
+            UserTokenCache = config.UserTokenCache;
+            if (UserTokenCache != null)
+            {
+                UserTokenCache.ClientId = ClientId;
+                UserTokenCache.ServiceBundle = ServiceBundle;
+            }
+
+            CreateRequestContext(Guid.Empty).Logger.Info(string.Format(CultureInfo.InvariantCulture,
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ClientApplicationBase.WithBuilders.cs]: >Info [](start = 52, length = 4)

should this be only in the verbose setting? I don't have a clear meaning on this myself, just wondering why you considered

File: src/Microsoft.Identity.Client/IPublicClientApplication.cs [hunk 2]
```diff
@@ -21,12 +21,12 @@
 // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-// THE SOFTWARE.
-//
+// THE SOFTWARE.//
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/IPublicClientApplication.cs]: >// [](start = 16, length = 2)

?

File: src/Microsoft.Identity.Client/ClientApplicationBase.WithBuilders.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ClientApplicationBase.WithBuilders.cs]: This is the level it was at before.  I can change it to make it verbose, but I'm trying not to make behavioral changes whenever possible.

---
In reply to: [247210395](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/780#discussion_r247210395) [](ancestors = 247210395)

File: src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 3]
```diff
@@ -106,12 +106,36 @@ public ConfidentialClientApplicationBuilder WithClientSecret(string clientSecret
...
+
+            if (Config.ClientCredential != null)
+            {
+                numSpecified++;
+            }
+
+            if (numSpecified > 1)
             {
                 throw new InvalidOperationException(
                     "ClientSecret and Certificate are mutually exclusive properties.  Only specify one.");
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs]: >ClientSecret and Certificate are mutually exclusive properties.  Only specify one." [](start = 21, length = 83)

aside: we have a string literals file for all error/exception messages

File: src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 4]
```diff
@@ -106,12 +106,36 @@ public ConfidentialClientApplicationBuilder WithClientSecret(string clientSecret
...
+            return this;
+        }
+
         /// <inheritdoc />
         internal override ApplicationConfiguration BuildConfiguration()
         {
             base.BuildConfiguration();
 
-            if (!string.IsNullOrWhiteSpace(Config.ClientSecret) && Config.Certificate != null)
+            int numSpecified = 0;
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs]: >numSpecified [](start = 16, length = 12)

nit: countOfCredentialTypesSpecified ?

File: src/Microsoft.Identity.Client/Http/HttpClientFactory.cs [hunk 5]
```diff
@@ -27,32 +27,31 @@
...
 
-    internal class HttpClientFactory : IHttpClientFactory
-    {
         // The HttpClient is a singleton per ClientApplication so that we don't have a process wide singleton.
         public const long MaxResponseContentBufferSizeInBytes = 1024*1024;
 
         public HttpClientFactory()
         {
-            var httpClient = new HttpClient(new HttpClientHandler() { UseDefaultCredentials = true })
+            _httpClient = new HttpClient(new HttpClientHandler() { UseDefaultCredentials = true })
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Http/HttpClientFactory.cs]: >_httpClient [](start = 12, length = 11)

The httpClient should be possible specified externally (e.g. on the config object). Not sure that is the PR, though?

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/Http/HttpClientFactory.cs]: We already have that:

public T WithHttpClientFactory(IMsalHttpClientFactory httpClientFactory)

This class is used if the developer doesn't specify their own HttpClientFactory

---
In reply to: [247265123](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/780#discussion_r247265123) [](ancestors = 247265123)

File: src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs (hunk 3 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs]: I'll add a TODO(migration) here to fix this and ensure we go through and clean all of these up.

---
In reply to: [247263898](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/780#discussion_r247263898) [](ancestors = 247263898)

**Pull request merged**